server/dive_tasks/tasks.py

```diff
@@ -545,3 +545,104 @@ def convert_images(self: Task, folderId):
             str(folderId),
             {"annotate": True},  # mark the parent folder as able to annotate.
         )
+
+
+@app.task(bind=True, acks_late=True, ignore_result=True)
+def extract_zip(self: Task, folderId: str, itemId: str):
+    context: dict = {}
+    gc: GirderClient = self.girder_client
+    manager: JobManager = patch_manager(self.job_manager)
+    if utils.check_canceled(self, context):
+        manager.updateStatus(JobStatus.CANCELED)
+        return
+
+    with tempfile.TemporaryDirectory() as _working_directory, suppress(utils.CanceledError):
+        _working_directory_path = Path(_working_directory)
+        item: GirderModel = gc.getItem(itemId)
+        file_name = str(_working_directory_path / item['name'])
+        manager.write(f'Fetching input from {itemId} to {file_name}...\n')
+        gc.downloadItem(itemId, _working_directory, item["name"])
+        top_level_folders = {}
+        with zipfile.ZipFile(file_name, 'r') as zipObj:
+            listOfFileNames = zipObj.namelist()
+            sum_file_size = sum([data.file_size for data in zipObj.filelist])
+            sum_compress_size = sum([data.compress_size for data in zipObj.filelist])
+            ratio = sum_file_size / sum_compress_size
+            if ratio > 600:
+                manager.write(
+                    f"Compression ratio is exceedingly high at {ratio}\n\
+                    Please contact an admin at viame-web@kitware.com if this is a valid zip file"
+                )
+                manager.updateStatus(JobStatus.ERROR)
+                return
+
+            for fileName in listOfFileNames:
+                # contains a sub folder
+                if fileName.endswith('/') and fileName.count('/') == 1:
+                    top_level_folders[fileName.replace('/', '')] = 'flat'
+                # sub folder has a meta.json so it is imported
+                if fileName.endswith(constants.MetaFileName):
+                    root_folder = fileName.replace(f'/{constants.MetaFileName}', '')
+                    top_level_folders[root_folder] = 'dataset'
+                if fileName.endswith('.zip'):
+                    manager.write("Nested Zip files are invalid\n")
+                    manager.updateStatus(JobStatus.ERROR)
+                    return
+                manager.write(f"Extracting: {fileName}\n")
+                zipObj.extract(fileName, f'{_working_directory}')
+        # remove the zip file
+        os.remove(file_name)
+        # validation of files in folder using dive/data
+        created_folder = gc.createFolder(
+            folderId,
+            constants.SourceFolderName,
+            reuseExisting=True,
+        )
+        gc.sendRestRequest(
+            "PUT",
+            f"/item/{str(item['_id'])}?folderId={str(created_folder['_id'])}",
+        )
+
+        if len(top_level_folders) > 0:
```

Comment:
I didn't know you could `len()` a dict, but it works!

Reply:
yeah I used to do `len(dict.keys())` until I saw that you could just `len(dict)` and it should get the number of keys.